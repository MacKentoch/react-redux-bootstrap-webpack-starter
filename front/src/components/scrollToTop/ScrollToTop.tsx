import React, { Fragment, ReactNode } from 'react';
import { useLocation } from 'react-router-dom';
import { useScrollToTopOnLocationChange } from './hooks/useScrollToTopOnLocationChange';
type OwnProps = {
  children: ReactNode;
};
type Props = OwnProps;

function ScrollToTop({ children }: Props) {
  const location = useLocation();
  useScrollToTopOnLocationChange(location);

  return <Fragment>{children}</Fragment>;
}

ScrollToTop.displayName = 'ScrollToTop';

export default ScrollToTop;
