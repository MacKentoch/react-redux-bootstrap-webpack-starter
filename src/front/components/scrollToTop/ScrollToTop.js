// @flow

// #region imports
import React, { Component } from 'react';
import {
  type Match,
  type Location,
  type RouterHistory,
} from 'react-router-dom';
import { withRouter } from 'react-router';
// #endregion

// #region flow types
type Props = {
  // react-router 4:
  match: Match,
  location: Location,
  history: RouterHistory,

  children: React.ReactNode,

  ...any,
};
type State = any;
// #endregion

class ScrollToTop extends Component<Props, State> {
  // #region lifecycle
  componentDidUpdate(prevProps: Props) {
    if (window) {
      const { location: prevLocation } = prevProps;
      const { location: nextLocation } = this.props;

      if (prevLocation !== nextLocation) {
        window.scrollTo(0, 0);
      }
    }
  }

  render() {
    const { children } = this.props;
    return children;
  }
  // #endregion
}

export default withRouter(ScrollToTop);
