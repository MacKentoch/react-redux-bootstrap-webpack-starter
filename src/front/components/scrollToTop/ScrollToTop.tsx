import React, { Fragment, useEffect, useRef } from 'react';
import { withRouter, RouteComponentProps } from 'react-router';

// #region types
type Props = {
  children: any,
} & RouteComponentProps;
// #endregion

function useScrollToTopOnLocationChange(location: any) {
  const prevLocation = useRef();

  useEffect(() => {
    prevLocation.current = location;
  }, []);

  useEffect(() => {
    if (prevLocation.current !== location) {
      window && window.scrollTo(0, 0);
      prevLocation.current = location;
    }
  }, [location]);
}

function ScrollToTop({ children, location }: Props) {
  useScrollToTopOnLocationChange(location);

  return <Fragment>{children}</Fragment>;
}

ScrollToTop.displayName = 'ScrollToTop';

export default withRouter(ScrollToTop);
