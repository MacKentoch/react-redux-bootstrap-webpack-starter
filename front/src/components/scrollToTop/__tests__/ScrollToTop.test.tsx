import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router';
import ScrollToTop from '../ScrollToTop';

describe('ScrollToTop component', () => {
  let rootElement: any = null;

  beforeEach(() => {
    rootElement = document.createElement('div');
    document.body.appendChild(rootElement);
  });

  afterEach(() => {
    rootElement && document.body.removeChild(rootElement);
    rootElement = null;
  });

  it('renders as expected', () => {
    const props = {
      brand: 'test',
      leftNavItemClick: jest.fn(),
      rightNavItemClick: jest.fn(),
      navModel: {
        brand: 'test',
        leftLinks: [
          {
            link: '/',
            label: 'home',
            viewName: 'home',
            onClick: jest.fn(),
          },
        ],
        rightLinks: [
          {
            link: '/',
            label: 'home',
            viewName: 'home',
            onClick: jest.fn(),
          },
        ],
      },
    };

    const { container } = render(
      <MemoryRouter>
        <ScrollToTop>
          <p>a child</p>
        </ScrollToTop>
      </MemoryRouter>,
      rootElement,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
