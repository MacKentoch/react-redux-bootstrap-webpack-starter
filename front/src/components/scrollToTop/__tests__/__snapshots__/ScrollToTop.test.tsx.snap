// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollToTop component renders as expected 1`] = `
<p>
  a child
</p>
`;
